onmt/translate/Beam.lua

```diff
@@ -1,190 +1,411 @@
+  tensors by specified indexes.
 
 Parameters:
 
-  * `wordLk`- probs of advancing from the last step (K x words)
-  * `featsLk`- probs of features at the last step (K x numfeatures x featsize)
-  * `attnOut`- attention at the last step
+  * `v` - tensor of size `(batchSize * beamSize, ...)` or a table containing such tensors
+  * `indexes` - a tensor of size `(batchSize, k)` specifying the desired indexes
+  * `beamSize` - beam size
+
+Returns: Indexed `(batchSize * k, ...)` tensor or a table containing such tensors
 
-Returns: true if beam search is complete.
 --]]
-function Beam:advance(wordLk, featsLk, attnOut)
+local function selectBeam(v, indexes, beamSize)
+  return onmt.utils.Tensor.recursiveApply(v, function (h)
+    local batchSize = indexes:size(1)
+    local k = indexes:size(2)
+    beamSize = beamSize or k
+    local sizes = {}
+    local ones = {}
+    for j = 2, #h:size() do
+        sizes[j - 1] = h:size(j)
+        ones[j - 1] = 1
+    end
+    return h:view(batchSize, beamSize, table.unpack(sizes))
+            :gather(2, indexes:view(batchSize, k, table.unpack(ones))
+                              :expand(batchSize, k, table.unpack(sizes)))
+            :view(batchSize * k, table.unpack(sizes))
+  end)
+end
 
-  -- The flattened scores.
-  local flatWordLk
+--[[Helper function. Recursively expand `(batchSize, ...)` tensors
+  to `(batchSize * beamSize, ...)` tensors.
+
+Parameters:
 
-  if #self.prevKs > 1 then
-    -- Sum the previous scores.
-    for k = 1, self.size do
-      wordLk[k]:add(self.scores[k])
+  * `v` - tensor of size `(batchSize, ...)` or a table containing such tensors
+  * `beamSize` - beam size
+
+Returns: Expanded `(batchSize * beamSize, ...)` tensor or a table containing
+  such tensors
+
+--]]
+local function replicateBeam(v, beamSize)
+  return onmt.utils.Tensor.recursiveApply(v, function (h)
+    local batchSize = h:size(1)
+    local sizes = {}
+    for j = 2, #h:size() do
+        sizes[j - 1] = h:size(j)
     end
-    flatWordLk = wordLk:view(-1)
+    return h:contiguous():view(batchSize, 1, table.unpack(sizes))
+                  :expand(batchSize, beamSize, table.unpack(sizes)):contiguous()
+                  :view(batchSize * beamSize, table.unpack(sizes))
+  end)
+end
+
+
+--[[Constructor.
+
+Parameters:
+
+  * `token` - tensor of size `(batchSize, vocabSize)` or `(batchSize * beamSize, vocabSize)`, or a list of such tensors.
+  * `state` - an iteratable object, where the contained tensors should have the same first dimension as `token`.
+  * `remaining` - remaining batch size. [`token:size(1)`]
+
+--]]
+function Beam:__init(token, state, remaining)
+  self._remaining = remaining or token:size(1)
+
+  if torch.type(token) == 'table' then
+    self._tokens = token
   else
-    flatWordLk = wordLk[1]:view(-1)
+    self._tokens = { token }
   end
+  self._state = state
 
+  self._scores = localize(torch.zeros(self._remaining), self._tokens[1])
```

Comment:
It seems to do the same as [`typeAs`](https://github.com/torch/torch7/blob/master/doc/tensor.md#tensor-typeastensor).